### ISSUE TITLE
Option to generate JSON schemas flattened

### ISSUE THREAD

**Issue by Dogacel** [line/armeria#4743]
Currently, we generate JSON Schemas greedily. If the field is another object, we check whether we have created an instance of that yet, if not we do it in place. Instead, we can add an option to generate them flat, meaning every object will be a reference to a `definition` using `$ref`.

I believe for the doc service schema flattening doesn't make sense because we have a schema for each request type. Therefore having this nested, greedy structure makes seeing the whole object shape much easier. But in OpenAPI spec, all endpoint definitions are grouped together, meaning having a flat structure helps with readability. Also having a `$ref` to an inner field of another definition is confusing.

For example, https://github.com/iheartradio/play-swagger generates those definitions flat as I have described as well.

**Comment by trustin** [comment]
Sounds like a good idea to me. At the moment, `specification.json` is once generated and cached IIRC. What would be the best way to implement this feature?

**Comment by Dogacel** [comment]
@trustin Yes, we do cache it and serve from the path `/schemas.json`. But in the use-case I state above, OpenAPI won't use the schema definitions from this endpoint. This endpoint is primarily going to be used by `DocService` for the autocompletion of request types. OpenAPI can invoke the `JsonSchemaGenerator` methods with the specified option.

I.e. this method can change to support the flat option

https://github.com/line/armeria/blob/72de6ff3ce1f732eebf2f325d0c54863fc7a0a5d/core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java#L98

Of course we can slightly refactor the code to change the entrypoint (service, method, field descriptor etc.)
